gateway/api/user/user.go

```diff
@@ -581,6 +581,177 @@ func ListAllGroups(c *gin.Context) {
 	c.JSON(http.StatusOK, groupsList)
 }
 
+// CreateGroup
+//
+//	@Summary		Create User Group
+//	@Description	Create a new user group
+//	@Tags			User Management
+//	@Accept			json
+//	@Produce		json
+//	@Param			request	body	openapi.UserGroup	true	"Group object"
+//	@Success		201		{object}	openapi.UserGroup	"{"name": "string"}"
+//	@Failure		400		{object}	openapi.HTTPError	"{"message": "error message"}"
+//	@Failure		409		{object}	openapi.HTTPError	"{"message": "group already exists"}"
+//	@Failure		500		{object}	openapi.HTTPError	"{"message": "server error"}"
+//	@Router			/users/groups [post]
+func CreateGroup(c *gin.Context) {
+	ctx := storagev2.ParseContext(c)
+
+	var req openapi.UserGroup
+
+	if err := c.ShouldBindJSON(&req); err != nil {
+		c.JSON(http.StatusBadRequest, gin.H{"message": err.Error()})
+		return
+	}
+
+	// Check if group already exists
+	userGroups, err := models.GetUserGroupsByOrgID(ctx.OrgID)
+	if err != nil {
+		log.Errorf("failed getting org groups, err=%v", err)
+		sentry.CaptureException(err)
+		c.JSON(http.StatusInternalServerError, gin.H{"message": "failed getting org groups"})
+		return
+	}
+
+	for _, ug := range userGroups {
+		if ug.Name == req.Name {
+			c.JSON(http.StatusConflict, gin.H{"message": fmt.Sprintf("group %s already exists", req.Name)})
+			return
+		}
+	}
```

Comment:
You're obtaining all user groups to validate if a single group exists. You could make this more simpler by removing the constraint in the underline function,  create the group and catch the unique constraint error gorm.ErrDuplicatedKey. See this function for more details: `UpsertGuardRailRuleWithConnections`

Reply:
I implemented the suggested improvement to make the code more efficient.

It wasn't possible to use the same duplicate key error pattern (gorm.ErrDuplicatedKey) as in UpsertGuardRailRuleWithConnections, because the user_groups table doesn't have a unique constraint for the (org_id, name) combination. However, the implemented approach is equally efficient and has the same behavior.